feat: migrate from Netlify CMS to Decap CMS and update dependencies

biru-codeastromer opened the pull request:
Fixes #1, fixes #78

### Description

This pull request refactors the current project setup by migrating from Netlify CMS to Decap CMS. The migration enhances CMS functionality and aligns with Decap CMS's rebranding and new features. Additionally, this PR modernizes the code by updating dependencies, addressing issue #78, and improving the overall project setup.

### Changes made:

1.Replaced the netlify-cms-app dependency with decap-cms in the package.json.
2.Switched from gatsby-plugin-netlify-cms to gatsby-plugin-decap-cms in gatsby-config.js.
3.Updated config.yml to ensure compatibility with Decap CMS's backend and media folder settings.
4.Modernized the project by updating outdated dependencies to improve compatibility and security.

krisstern commented:
Hi @biru-codeastromer please refer to this doc https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests#linking-a-pull-request-to-an-issue on how to use keywords to properly link an issue. I have fixed it up for you previously. But you may want to do so properly moving forwards.

krisstern commented:
Also you may want to run the build locally to make sure there are no issues before pushing code to the repo. I am current seeing the following in the logs at https://github.com/jenkins-infra/stories/pull/79/checks?check_run_id=34477938558:

```console
[2024-12-16T14:49:32.936Z] + npm run lint
[2024-12-16T14:49:33.238Z] 
[2024-12-16T14:49:33.238Z] > stories-jenkins.io@1.0.0 lint
[2024-12-16T14:49:33.238Z] > eslint .
[2024-12-16T14:49:33.238Z] 
[2024-12-16T14:49:34.942Z] Browserslist: caniuse-lite is outdated. Please run:
[2024-12-16T14:49:34.942Z]   npx browserslist@latest --update-db
[2024-12-16T14:49:34.943Z]   Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
[2024-12-16T14:49:35.444Z] 
[2024-12-16T14:49:35.445Z] /home/jenkins/agent/workspace/Websites_stories_PR-79/gatsby-config.js
[2024-12-16T14:49:35.445Z]   25:11  error  Trailing spaces not allowed  no-trailing-spaces
[2024-12-16T14:49:35.445Z] 
[2024-12-16T14:49:35.445Z] /home/jenkins/agent/workspace/Websites_stories_PR-79/src/cms/cms.js
[2024-12-16T14:49:35.445Z]    1:17  error  Unable to resolve path to module 'decap-cms-app'  import/no-unresolved
[2024-12-16T14:49:35.445Z]    3:20  error  Missing file extension for "../layout"            import/extensions
[2024-12-16T14:49:35.445Z]    3:20  error  Unable to resolve path to module '../layout'      import/no-unresolved
[2024-12-16T14:49:35.445Z]   13:28  error  There should be no space after '{'                object-curly-spacing
[2024-12-16T14:49:35.445Z]   13:56  error  There should be no space before '}'               object-curly-spacing
[2024-12-16T14:49:35.445Z]   23:35  error  Unexpected function expression                    prefer-arrow-callback
[2024-12-16T14:49:35.446Z]   28:10  error  Missing semicolon                                 semi
[2024-12-16T14:49:35.446Z]   35:5   error  Unexpected console statement                      no-console
[2024-12-16T14:49:35.446Z] 
[2024-12-16T14:49:35.446Z] ✖ 9 problems (9 errors, 0 warnings)
[2024-12-16T14:49:35.446Z]   5 errors and 0 warnings potentially fixable with the `--fix` option.
[2024-12-16T14:49:35.446Z] 
```

krisstern reviewed (changes requested):
Please update the file name `src/cms/cms.js` to `src/cms/index.js`, and update all references to this path accordingly.

biru-codeastromer commented:
> Hi @biru-codeastromer please refer to this doc https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests#linking-a-pull-request-to-an-issue on how to use keywords to properly link an issue. I have fixed it up for you previously. But you may want to do so properly moving forwards.

Hi @krisstern Sir,

Thank you for pointing that out! I appreciate you fixing the issue previously. I’ll make sure to follow the guidelines provided in the doc moving forward and use the appropriate keywords to link issues and pull requests properly. I’ll refer to the resource to ensure I’m following the best practices.

Thanks for your help!

biru-codeastromer commented:
> Please update the file name `src/cms/cms.js` to `src/cms/index.js`, and update all references to this path accordingly.

done sir

biru-codeastromer commented:
> Also you may want to run the build locally to make sure there are no issues before pushing code to the repo. I am current seeing the following in the logs at https://github.com/jenkins-infra/stories/pull/79/checks?check_run_id=34477938558:


Hi @krisstern,

I’ve completed fixing the previous errors, but I am still encountering issues with the layout and Decap CMS integration. It seems like the layout module is not being found in the src folder, which is causing problems with the build. Additionally, the decap-cms-app module also isn’t resolving properly.

Interestingly, the same issue I saw in the template provided by the zbynek  Sir comment on #1 , where the problem also occurred in the build of that repo.
The issue arises because decap-cms-app@3.4.0 has a peer dependency requirement for react@^18.2.0, while our repo is using react@17.0.2. This results in a conflict that prevents the installation from proceeding.

To resolve the npm install decap-cms-app issue, I used the yarn method as an alternative to npm. While this did successfully solve the dependency conflict, I’ve encountered a new issue after deploying the changes. Specifically, the images are no longer showing up on the deployed site.

It seems like the change may have affected how the assets are being referenced or loaded. I’m currently looking into this issue.

After resolving all the other errors, these 3 remaining issues persist in the logs:
/Users/birajitsaikia/Documents/decapcms/stories/src/cms/index.js
  birajitsaikia@Birajits-MacBook-Pro stories % npm run lint
  1:17  error  Unable to resolve path to module 'decap-cms-app'  import/no-unresolved
  3:20  error  Missing file extension for "../layout"            import/extensions
  3:20  error  Unable to resolve path to module '../layout'      import/no-unresolved

krisstern commented:
Please note we also have this issue #78 for moderning dependecies, maybe you can use the same PR to resolve both issues.

biru-codeastromer commented:
> Please note we also have this issue #78 for modernizing dependecies, maybe you can use the same PR to resolve both issues.

Okay Sir looking into it !

biru-codeastromer commented:
<img width="1024" alt="Screenshot 2024-12-17 at 4 34 20 AM" src="https://github.com/user-attachments/assets/d1079590-e1da-468f-87db-96b00550cc93" />

I corrected the required errors and did it successfully the building .Still some ci checks  failed and will now look onto it.

biru-codeastromer commented:
When I try to correct the ci checks, it going two ways:

1. If i keep the name as gatsby.js and make some other changes i get the build log error as : 
`[2024-12-16T23:40:18.122Z] + npm run build
[2024-12-16T23:40:18.323Z] 
[2024-12-16T23:40:18.323Z] > stories-jenkins.io@1.0.0 build
[2024-12-16T23:40:18.323Z] > gatsby build
[2024-12-16T23:40:18.323Z] 
[2024-12-16T23:40:21.429Z] success compile gatsby files - 1.255s
[2024-12-16T23:40:21.430Z] error We encountered an error while trying to load your site's gatsby-config. Please fix the error and try again.
[2024-12-16T23:40:21.430Z] 
[2024-12-16T23:40:21.430Z] 
[2024-12-16T23:40:21.430Z]   ReferenceError: module is not defined in ES module scope
[2024-12-16T23:40:21.430Z]   This file is being treated as an ES module because it has a '.js' file extensi  on and '/home/jenkins/agent/workspace/Websites_stories_PR-79/package.json' con  tains "type": "module". To treat it as a CommonJS script, rename it to use the   '.cjs' file extension.
[2024-12-16T23:40:21.430Z]   
[2024-12-16T23:40:21.430Z]   - gatsby-config.js:1 
[2024-12-16T23:40:21.430Z]     file:///home/jenkins/agent/workspace/Websites_stories_PR-79/gatsby-config.js    :1:1
[2024-12-16T23:40:21.430Z]   
[2024-12-16T23:40:21.430Z]   - module_job:234 ModuleJob.run
[2024-12-16T23:40:21.430Z]     node:internal/modules/esm/module_job:234:25
[2024-12-16T23:40:21.430Z]   
[2024-12-16T23:40:21.430Z]   - loader:473 async ModuleLoader.import
[2024-12-16T23:40:21.430Z]     node:internal/modules/esm/loader:473:24
[2024-12-16T23:40:21.431Z]   
[2024-12-16T23:40:21.431Z]   - get-config-file.js:35 async attemptImport
[2024-12-16T23:40:21.431Z]     [Websites_stories_PR-79]/[gatsby]/dist/bootstrap/get-config-file.js:35:26
[2024-12-16T23:40:21.431Z]   
[2024-12-16T23:40:21.431Z]   - get-config-file.js:64 async attemptImportUncompiled
[2024-12-16T23:40:21.431Z]     [Websites_stories_PR-79]/[gatsby]/dist/bootstrap/get-config-file.js:64:24
[2024-12-16T23:40:21.431Z]   
[2024-12-16T23:40:21.431Z]   - get-config-file.js:19 async getConfigFile
[2024-12-16T23:40:21.431Z]     [Websites_stories_PR-79]/[gatsby]/dist/bootstrap/get-config-file.js:19:28
[2024-12-16T23:40:21.431Z]   
[2024-12-16T23:40:21.431Z]   - index.js:22 async loadConfig
[2024-12-16T23:40:21.431Z]     [Websites_stories_PR-79]/[gatsby]/dist/bootstrap/load-config/index.js:22:7
[2024-12-16T23:40:21.431Z]   
[2024-12-16T23:40:21.431Z]   - initialize.js:139 async initialize
[2024-12-16T23:40:21.431Z]     [Websites_stories_PR-79]/[gatsby]/dist/services/initialize.js:139:18
[2024-12-16T23:40:21.432Z]   
[2024-12-16T23:40:21.432Z]   - index.js:27 async bootstrap
[2024-12-16T23:40:21.432Z]     [Websites_stories_PR-79]/[gatsby]/dist/bootstrap/index.js:27:9
[2024-12-16T23:40:21.432Z]   
[2024-12-16T23:40:21.432Z]   - build.js:82 async build
[2024-12-16T23:40:21.432Z]     [Websites_stories_PR-79]/[gatsby]/dist/commands/build.js:82:7
[2024-12-16T23:40:21.432Z]   
[2024-12-16T23:40:21.432Z] 
[2024-12-16T23:40:21.432Z] not finished load gatsby config - 0.055s
`

2.And when i again changed to gatsby.cjs and make other required changes ,I get this : 
`[2024-12-16T23:30:23.459Z] + npm run build
[2024-12-16T23:30:23.761Z] 
[2024-12-16T23:30:23.761Z] > stories-jenkins.io@1.0.0 build
[2024-12-16T23:30:23.761Z] > gatsby build
[2024-12-16T23:30:23.761Z] 
[2024-12-16T23:30:26.968Z] success compile gatsby files - 1.407s
[2024-12-16T23:30:26.968Z] error It looks like you were trying to add the config file? Please rename "gatsby-config.cjs" to "gatsby-config.js"
[2024-12-16T23:30:26.968Z] 
[2024-12-16T23:30:26.968Z] 
[2024-12-16T23:30:26.968Z]   Error: Cannot find package '/home/jenkins/agent/workspace/Websites_stories_PR-  79/gatsby-config'
[2024-12-16T23:30:26.968Z]   
[2024-12-16T23:30:26.969Z]   - get-config-file.js:80 attemptImportUncompiled
[2024-12-16T23:30:26.969Z]     [Websites_stories_PR-79]/[gatsby]/dist/bootstrap/get-config-file.js:80:17
[2024-12-16T23:30:26.969Z]   
[2024-12-16T23:30:26.969Z]   - task_queues:95 process.processTicksAndRejections
[2024-12-16T23:30:26.969Z]     node:internal/process/task_queues:95:5
[2024-12-16T23:30:26.969Z]   
[2024-12-16T23:30:26.969Z]   - get-config-file.js:19 async getConfigFile
[2024-12-16T23:30:26.969Z]     [Websites_stories_PR-79]/[gatsby]/dist/bootstrap/get-config-file.js:19:28
[2024-12-16T23:30:26.969Z]   
[2024-12-16T23:30:26.969Z]   - index.js:22 async loadConfig
[2024-12-16T23:30:26.969Z]     [Websites_stories_PR-79]/[gatsby]/dist/bootstrap/load-config/index.js:22:7
[2024-12-16T23:30:26.969Z]   
[2024-12-16T23:30:26.969Z]   - initialize.js:139 async initialize
[2024-12-16T23:30:26.970Z]     [Websites_stories_PR-79]/[gatsby]/dist/services/initialize.js:139:18
[2024-12-16T23:30:26.970Z]   
[2024-12-16T23:30:26.970Z]   - index.js:27 async bootstrap
[2024-12-16T23:30:26.970Z]     [Websites_stories_PR-79]/[gatsby]/dist/bootstrap/index.js:27:9
[2024-12-16T23:30:26.970Z]   
[2024-12-16T23:30:26.970Z]   - build.js:82 async build
[2024-12-16T23:30:26.970Z]     [Websites_stories_PR-79]/[gatsby]/dist/commands/build.js:82:7
[2024-12-16T23:30:26.970Z]   
[2024-12-16T23:30:26.970Z] 
[2024-12-16T23:30:26.970Z] not finished load gatsby config - 0.047s`



this is happening vice-versa whatever suggestion i do it goes either way.

krisstern commented:
For upgrading from React 17 to 18 there are additional steps like they outline here https://react.dev/blog/2022/03/08/react-18-upgrade-guide. So @biru-codeastromer you may want to do some more background research instead of going by trial-and-error.

krisstern commented:
@biru-codeastromer Could you please remove the `.DS_Store` you have added and add an entry for `.DS_Store` in the `.gitignore` file?

krisstern commented:
Also, could you please try running `npm audit fix` as well for this PR?

krisstern commented:
I also see the following in the stack trace currently:
```console
Error: Cannot create as TypeComposer the following value: 
    GraphQLScalarType({ name: "Date", description: "A date string, such as 2007-12-03, compliant with the ISO 8601
   standard for representation of dates and times using the Gregorian calendar.", specifiedByURL: undefined, seria
  lize: [function String], parseValue: [function String], parseLiteral: [function parseLiteral], extensions: {  },
   astNode: undefined, extensionASTNodes: [] }).
```

krisstern commented:
Could you also check if we will need to upgrade from `gatsby-plugin-netlify` to `gatsby-plugin-decap-cms` in the dependencies? Otherwise it may not work.

krisstern commented:
Is there a reason why I am seeing
```json
"react": "^18.0.0",
    "react-dom": "^18.0.0",
```
in the `package.json` instead of the latest v18 version? Could you please update this as well?

biru-codeastromer commented:
> For upgrading from React 17 to 18 there are additional steps like they outline here https://react.dev/blog/2022/03/08/react-18-upgrade-guide. So @biru-codeastromer you may want to do some more background research instead of going by trial-and-error.


Hi Sir @krisstern ,
Thank you for pointing this out! React 18 has installed successfully ,But I will review the React 18 Upgrade Guide and ensure all necessary steps are taken, such as reviewing breaking changes, enabling features like concurrent rendering, and updating components to align with React 18 best practices.

biru-codeastromer commented:
> @biru-codeastromer Could you please remove the `.DS_Store` you have added and add an entry for `.DS_Store` in the `.gitignore` file?

Sure Sir going to do it.

biru-codeastromer commented:
> Also, could you please try running `npm audit fix` as well for this PR?

I’ve run npm audit fix multiple times during this process, but it hasn’t resolved all vulnerabilities. I will review the remaining audit issues manually and see if further updates or alternate solutions are required.

biru-codeastromer commented:
> I also see the following in the stack trace currently:
> 
> ```
> Error: Cannot create as TypeComposer the following value: 
>     GraphQLScalarType({ name: "Date", description: "A date string, such as 2007-12-03, compliant with the ISO 8601
>    standard for representation of dates and times using the Gregorian calendar.", specifiedByURL: undefined, seria
>   lize: [function String], parseValue: [function String], parseLiteral: [function parseLiteral], extensions: {  },
>    astNode: undefined, extensionASTNodes: [] }).
> ```




Hi Sir @krisstern  ,regarding the GraphQL error, I have made several attempts to resolve the issue, but the same error persists. I am currently investigating the root cause more deeply, including examining plugin compatibility and schema handling.

Additionally, I’ll check the feasibility of replacing gatsby-plugin-netlify with gatsby-plugin-decap-cms to improve compatibility and align with updated project requirements.

Thank you for your guidance!

biru-codeastromer commented:
<img width="1098" alt="Screenshot 2024-12-18 at 7 30 50 PM" src="https://github.com/user-attachments/assets/f438fbd9-4372-429c-a2bd-14d441030823" />

I found this on documentation of gatsby. The error i am facing is mentioned ,It is said to be cause to multiple.I tried using resolutions tactics but same issue came up again .Should I use this " delete node_modules and re-install with npm install --legacy-peer-deps/yarn install" to tackle this Sir.

biru-codeastromer commented:
<img width="1512" alt="Screenshot 2024-12-18 at 9 12 31 PM" src="https://github.com/user-attachments/assets/89e6650f-6061-4c00-aeb3-83f633eb0f85" />
<img width="1512" alt="Screenshot 2024-12-18 at 9 12 18 PM" src="https://github.com/user-attachments/assets/459631d3-882c-42b4-a48c-eb3ba7a96728" />


Hi Sir  @krisstern  , I solved the issue without legacy peer deps.
###Changes Made

1.Renamed Scalar Type:

a.Changed the custom scalar type from Date to CustomDate in gatsby-node.js.
b.Updated schema customization to define CustomDate directly.

2.Removed Redundant Resolvers:

a.Removed the createResolvers function for the CustomDate scalar type as it's not required for scalar types.

3.Ensured Single Version of graphql:

a.Added resolutions and overrides in package.json to enforce the use of a single version of graphql (16.10.0).

biru-codeastromer commented:
> Remember to push your code changes here

I encountered a permission issue during the build process in our CI pipeline. The error message indicated that there was a permission denied error when trying to create a directory in the /opt/build path. Here are the details of the error:


`[2024-12-18T16:36:44.466Z] error Cannot copy from `.cache/.eslintrc.json` to `/opt/build/cache/cwd/.cache/.eslintrc.json`: Cannot create directory `/opt/build/cache/cwd/.cache`: EACCES: permission denied, mkdir '/opt/build'
[2024-12-18T16:36:44.466Z] 
[2024-12-18T16:36:44.466Z] 
[2024-12-18T16:36:44.466Z]   NestedError: Cannot copy from `.cache/.eslintrc.json` to `/opt/build/cache/cwd  /.cache/.eslintrc.json`: Cannot create directory `/opt/build/cache/cwd/.cache`  : EACCES: permission denied, mkdir '/opt/build'
[2024-12-18T16:36:44.467Z]   
[2024-12-18T16:36:44.467Z]   - index.js:274 pMap.concurrency.concurrency
[2024-12-18T16:36:44.467Z]     [adapters]/[cpy]/index.js:274:12
[2024-12-18T16:36:44.467Z]   
[2024-12-18T16:36:44.467Z]   - index.js:141 
[2024-12-18T16:36:44.467Z]     [adapters]/[p-map]/index.js:141:20
[2024-12-18T16:36:44.467Z]   
[2024-12-18T16:36:44.467Z]   - Caused By: NestedError: Cannot create directory `/opt/build/cache/cwd/.cache    `: EACCES: permission denied, mkdir '/opt/build'
[2024-12-18T16:36:44.467Z]   
[2024-12-18T16:36:44.467Z]   - fs.js:61 
[2024-12-18T16:36:44.467Z]     [adapters]/[cp-file]/fs.js:61:8
[2024-12-18T16:36:44.467Z]   
[2024-12-18T16:36:44.467Z]   - index.js:15 async cpFileAsync
[2024-12-18T16:36:44.467Z]     [adapters]/[cp-file]/index.js:15:2
[2024-12-18T16:36:44.467Z]   
[2024-12-18T16:36:44.467Z]   - index.js:269 async pMap.concurrency.concurrency
[2024-12-18T16:36:44.467Z]     [adapters]/[cpy]/index.js:269:6
[2024-12-18T16:36:44.467Z]   
[2024-12-18T16:36:44.467Z]   - index.js:141 
[2024-12-18T16:36:44.467Z]     [adapters]/[p-map]/index.js:141:20
[2024-12-18T16:36:44.467Z]   
[2024-12-18T16:36:44.467Z]   - Caused By: Error: EACCES: permission denied, mkdir '/opt/build'`

krisstern commented:
It looks like the build step (which is the only one you have control of) has been successful, and the problem looks like a deployment issue as discussed for example in https://community.cloudflare.com/t/gatsby-pages-deployment-error/558278/5. Let me follow this up with the Jenkins Infra team shortly.

krisstern commented:
I have opened an issue with the Infra team helpdesk at https://github.com/jenkins-infra/helpdesk/issues/4461. But please do not comment on that issue. In case of anything you would like to voice out let me know and I will liaise with the Infra team via that ticket.

biru-codeastromer commented:
> It looks like the build step (which is the only one you have control of) has been successful, and the problem looks like a deployment issue as discussed for example in https://community.cloudflare.com/t/gatsby-pages-deployment-error/558278/5. Let me follow this up with the Jenkins Infra team shortly.

Thank you Sir for clarifying that the build step completed successfully and for identifying the deployment issue. I appreciate the reference to the related Cloudflare community discussion and your initiative to follow up with the Jenkins Infra team.

biru-codeastromer commented:
Hi Sir,

To resolve the permission issue during the build process, I propose using a custom cache directory. This involves setting environment variables and updating the build command to specify a directory with appropriate permissions:

1.Set Environment Variables
2.Update Build Command

Please let me know if you agree with this approach/thought or if there are any additional considerations.

krisstern commented:
Hi @biru-codeastromer The Infra team will come up with a solution shortly. Unless they ask for our opinion, we do not need to provide any feedback in this stage.

biru-codeastromer commented:
> Hi @biru-codeastromer The Infra team will come up with a solution shortly. Unless they ask for our opinion, we do not need to provide any feedback at this stage.

Thank You Sir for the clarification .

krisstern commented:
Also, this is not a trivial change. Once we update the configuration it will affect the entire repo moving forwards, so it is better to let the Infra team step in and do some magic for us. However, since we are officially in the winter break, so the change may not come as soon as it normally does.

biru-codeastromer commented:
> Also, this is not a trivial change. Once we update the configuration it will affect the entire repo moving forwards, so it is better to let the Infra team step in and do some magic for us. However, since we are officially in the winter break, so the change may not come as soon as it normally does.

Okay Sir, thank you for the details. Sure, till then I will continue solving other issues and make other contributions.

krisstern commented:
Hi @biru-codeastromer could you please take some time to resolve the conclicts in this PR?

biru-codeastromer commented:
> Hi @biru-codeastromer could you please take some time to resolve the conclicts in this PR?

Sure Sir @krisstern will be doing It !

biru-codeastromer commented:
Hi Sir @krisstern  I have resolved the  merge conflicts in the PR. 
Thanks!

biru-codeastromer commented:
> Hi @biru-codeastromer could you please take some time to resolve the conclicts in this PR?

I solved it , but due to regular changes in main it keeps getting conflicts ,
do you think Sir , I should wait for the PR's #80 and #84 to get merged then finally solved the conflicts at once . Or should I resolved it regularly . Please let me know

biru-codeastromer commented:
> I solved it , but due to regular changes in main it keeps getting conflicts , do you think Sir , I should wait for the PR's #80 and #84 to get merged then finally solved the conflicts at once . Or should I resolved it regularly . Please let me know !



Also I will follow your method what you told me in an issue regarding merging main

biru-codeastromer commented:
I will be resolving the conflicts today again and keep the updates here !

biru-codeastromer commented:
In the meantime , I have resolved the conflicts here . Thanks !

krisstern commented:
@dduportal Any updates on https://github.com/jenkins-infra/helpdesk/issues/4461 which is a blocker to the review of this PR?

dduportal commented:
> @dduportal Any updates on [jenkins-infra/helpdesk#4461](https://github.com/jenkins-infra/helpdesk/issues/4461) which is a blocker to the review of this PR?

See https://github.com/jenkins-infra/helpdesk/issues/4461#issuecomment-2668686089

krisstern commented:
@biru-codeastromer Could you please resolve the merge conflicts here?

biru-codeastromer commented:
Sure , I will be doing now .

biru-codeastromer commented:
I have resolved the conflicts . Although , I saw duplicates in package-lock.json after resolving so I deleted it and ran `npm i` again and made a new push to remove them . Now , I think it is fine . Let me know how you see it . Thanks !

dduportal commented:
> I have resolved the conflicts . Although , I saw duplicates in `package-lock.json`file after resolving so I deleted it and ran `npm i` again and made a new push to remove them . Now , I think it is fine . Let me know how you see it . Thanks !

As you can see in the checks (look https://github.com/jenkins-infra/stories/pull/79/checks?check_run_id=38907898330 or in ci.jenkins.io which generated this check) , the build is failing during the `npm run build` stage. It's an interesting problem to solve for you: does it run without error on your machine? If yes, what are the environment differences with the agents of ci.jenkins.io which are the reference?

Note: the build of preview is private (requires a VPN) but has the same error.

biru-codeastromer commented:
I didn't do `gatsby clean` which was needed before pushing and now I did removed `package-lock.json` again and then `npm i` followed by `gatsby clean && gatsby build` and it ran successfully in my machine 

<img width="1121" alt="Screenshot 2025-03-17 at 11 47 35 PM" src="https://github.com/user-attachments/assets/adaa34f5-6943-449c-83a2-c957797ec7ee" />


And now the error which you told is solved in the checks , although again now we come back to the original ci checks fails which is the "permission denied error when trying to create a directory in the /opt/build path."

biru-codeastromer commented:
also Thanks a lot for pointing the previous error .

dduportal commented:
> I didn't do `gatsby clean` which was needed before pushing and now I did removed `package-lock.json` again and then `npm i` followed by `gatsby clean && gatsby build` and it ran successfully in my machine
> 
> <img alt="Screenshot 2025-03-17 at 11 47 35 PM" width="1121" src="https://private-user-images.githubusercontent.com/184142102/423614725-adaa34f5-6943-449c-83a2-c957797ec7ee.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMzYxNjQsIm5iZiI6MTc0MjIzNTg2NCwicGF0aCI6Ii8xODQxNDIxMDIvNDIzNjE0NzI1LWFkYWEzNGY1LTY5NDMtNDQ5Yy04M2EyLWM5NTc3OTdlYzdlZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxN1QxODI0MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yODVkNDQ1YzEwYmQwMjIyMjg0ODYxOTJmYzRiMTNjOWQ0NjRjMzQwODlkZTM5OTA3YTA4ZDhkYThkOWY0OGZiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7doa-j4grwxw2-NUBzt8WETqfRAXzbqpGo3NdD4Jcg4">
> And now the error which you told is solved in the checks , although again now we come back to the original ci checks fails which is the "permission denied error when trying to create a directory in the /opt/build path."
> 
> https://github.com/jenkins-infra/stories/pull/79/checks?check_run_id=38911783996

- unrelated to your issue but important for this PR: « npm i » should not be default as it changes the content of package-lock.json. Good practise for reproductible builds is to use « npm ci » instead (which installs dependencies from the lock file instead of trying to regenerate it).
- The permission denied is what i explained on another issue: there are no good reason to write to this directory. Please check gastby documentation to find the parameter which allows to override this path

krisstern commented on src/components/Seo.jsx:
```diff
@@ -88,4 +83,4 @@ Seo.propTypes = {
 };
 
-export default Seo;
+export default Seo;
```
```suggestion
export default Seo;

```

biru-codeastromer commented:
I am back ! 
Now I will focus on and implement what Sir Damien instructed, as I have recently completed my draft proposal and submitted it for review. Now I can concentrate on this as well.

biru-codeastromer commented:
> * unrelated to your issue but important for this PR: « npm i » should not be default as it changes the content of package-lock.json. Good practise for reproductible builds is to use « npm ci » instead (which installs dependencies from the lock file instead of trying to regenerate it).
> * The permission denied is what i explained on another issue: there are no good reason to write to this directory. Please check gastby documentation to find the parameter which allows to override this path


Before one week when you said about the gatsby doc, I did some searches to find some valid documentation regarding this . and I thought of a solution , I want to know from you as well as Kris Sir about my approach how is it.

First I set `GATSBY_CACHE_DIR=./.custom-cache` to override Gatsby's default `.cache` location and ensure that cache is written to the project workspace (not `/opt/build`)

Then create cache directory in ci scripts and importantly update build commands to use custom path

then in `gatsby-config.js`  I will add cache-preservation flags 
``` 
flags: {
  PRESERVE_WEBPACK_CACHE: true, 
  FAST_DEV: true  
} 
```
and last update clean script to remove the custom cache..


Let me know how you think about my approach for overriding the path

dduportal commented:
> > * unrelated to your issue but important for this PR: « npm i » should not be default as it changes the content of package-lock.json. Good practise for reproductible builds is to use « npm ci » instead (which installs dependencies from the lock file instead of trying to regenerate it).
> > * The permission denied is what i explained on another issue: there are no good reason to write to this directory. Please check gastby documentation to find the parameter which allows to override this path
> 
> @dduportal
> 
> Before one week when you said about the gatsby doc, I did some searches to find some valid documentation regarding this . and I thought of a solution , I want to know from you as well as Kris Sir about my approach how is it.
> 
> First I set `GATSBY_CACHE_DIR=./.custom-cache` to override Gatsby's default `.cache` location and ensure that cache is written to the project workspace (not `/opt/build`)
> 
> Then create cache directory in ci scripts and importantly update build commands to use custom path
> 
> then in `gatsby-config.js` I will add cache-preservation flags
> 
> ```
> flags: {
>   PRESERVE_WEBPACK_CACHE: true, 
>   FAST_DEV: true  
> } 
> ```
> 
> and last update clean script to remove the custom cache..
> 
> Let me know how you think about my approach for overriding the path

looks like you're in the right direction to me!

biru-codeastromer commented:
Thanks I will try to implement it and see how it works in our PR.

krisstern commented:
Sounds good @biru-codeastromer do keep us posted!

krisstern commented:
@dduportal it does look like it is a setup issue which can only be diagnose by someone with admin access, i.e. someone from the Infra team.

dduportal commented:
> @dduportal it does look like it is a setup issue which can only be diagnosed by someone with admin access, i.e. someone from the Infra team.

Good catch: as @biru-codeastromer is not a contributor of this repository, ci.jenkins.io consider the PR "untrusted" and uses the `Jenkinsfile` from the principal branch instead of the modified `Jenkinsfile` in the PR. It is a default security mechanism to limit risks of malicious code injection (even if there are other techniques).

@krisstern @MarkEWaite I believe we could grant @biru-codeastromer additiona permissions ("Writer"' is enough, maybe "Read/Triage" should be good enough but I'm not sure as I never tried) on this repository?

On the short term, I'm triggering a build as admin (a replay in which I'm copy and pasting the `Jenkinsfile`) to validate @biru-codeastromer changes

krisstern commented:
Hi @dduportal maybe instead of simply doing `mkdir '/opt/build'` in the build config, before this step we could try changing the ownership to the current user as group:

```bash
sudo chown -R root:$(whoami) /opt
sudo chmod -R 775 /opt
```

But I sensed possibly it would be much better we should use the `/var/www` or even `/home/ubuntu` directory for placing the build folder instead? This is because `/opt` should be reserved for dependencies only, not for a folder to collect the built files of a website to be served.

dduportal commented:
> > @dduportal it does look like it is a setup issue which can only be diagnosed by someone with admin access, i.e. someone from the Infra team.
> 
> Good catch: as @biru-codeastromer is not a contributor of this repository, ci.jenkins.io consider the PR "untrusted" and uses the `Jenkinsfile` from the principal branch instead of the modified `Jenkinsfile` in the PR. It is a default security mechanism to limit risks of malicious code injection (even if there are other techniques).
> 
> @krisstern @MarkEWaite I believe we could grant @biru-codeastromer additiona permissions ("Writer"' is enough, maybe "Read/Triage" should be good enough but I'm not sure as I never tried) on this repository?
> 
> On the short term, I'm triggering a build as admin (a replay in which I'm copy and pasting the `Jenkinsfile`) to validate @biru-codeastromer changes

https://ci.jenkins.io/job/Websites/job/stories/job/PR-79/31/console

krisstern commented:
@dduportal No, I don't think we should grant @biru-codeastromer any access other than the one he has now.

krisstern commented:
Wait, let me see if there is somehthing I can do here

dduportal commented:
> Hi @dduportal maybe instead of simply doing `mkdir '/opt/build'` in the build config, before this step we could try changing the ownership to the current user as group:
> 
> ```shell
> sudo chown -R root:$(whoami) /opt
> sudo chmod -R 775 /opt
> ```
> 
> But I sensed possibly it would be much better we should use the `/var/www` or even `/home/ubuntu` directory for placing the build folder instead? This is because `/opt` should be reserved for dependencies only, not for a folder to collect the built files of a website to be served.

`/opt` or `/var` should be written by the agent's `jenkins` user, it is a safety measure. There are technique to do it if you have malicious intent, but we should not (we have to consider them readonly at all costs as we might enforce this on infra side). The local workspace is mandatory to use.

I believe the latest changes are in the correct direction though (see my message above)

krisstern commented:
> @dduportal No, I don't think we should grant @biru-codeastromer any access other than the one he has now.

I personally have experienced some issues from a former GSoC contributor whom we have granted admin access (or write access) and it was not very pleasant. I think this right should be earned eventually and not be given freely.

krisstern commented:
@dduportal do you think it would work if I modify the Jenkinsfile in this PR and push to the fork used?

dduportal commented:
> Wait, let me see if there is somehthing I can do here

Thanks for the commit to trigger a build. Alas it does not work: one maintainer/admin must "Replay" and copy and paste the modified `Jenkinsfile` in the field. That is how the security mechanism works.

I've retriggered by build 31: https://ci.jenkins.io/job/Websites/job/stories/job/PR-79/33/console has it was not completed

dduportal commented:
> @dduportal do you think it would work if I modify the Jenkinsfile in this PR and push to the fork used?

Mmmh good question: will the PR will be considered trusted if you co-author on commit on it 🤔 . Let's wait for build 33 to finish and we can try: that would give you more autonomy if it works

krisstern commented:
> Mmmh good question: will the PR will be considered trusted if you co-author on commit on it 🤔 . Let's wait for build 33 to finish and we can try: that would give you more autonomy if it works

Sure, sounds good!

dduportal commented:
OK, so I see `> GATSBY_CACHE_DIR=./.gatsby-cache GATSBY_INTERNAL_CACHE_DIR=./.cache gatsby build` as expected in the build 33 with the replay.

Alas, it sill fails with `error Cannot copy from `.cache/.eslintrc.json` to `/opt/build/cache/cwd/.cache/.eslintrc.json`: Cannot create directory `/opt/build/cache/cwd/.cache`: EACCES: permission denied, mkdir '/opt/build'` during a later phase in the `npm run ci:build`.

There might be other components at stake

@krisstern you are welcome to try co-authoring a change in the `Jenkinsfile` to see if you have the same result

krisstern commented:
Sure @dduportal just pushed some more changes

krisstern commented:
@dduportal I don't think the setting `${WORKSPACE}/` works for us for the caching... Should we try hardcoding this param to something like the home directory or `/var/www` or /www/wwwroot`?

krisstern commented:
I changed `${WORKSPACE}` to `${env.WORKSPACE}` after checking some documentation, let's see if this makes a difference.

dduportal commented:
> I changed `${WORKSPACE}` to `${env.WORKSPACE}` after checking some documentation, let's see if this makes a difference.

good catch. However, it does not look like that your commit are considered trusted: if I click "Replay" on the build 38 (triggered by https://github.com/jenkins-infra/stories/pull/79/commits/3a357da057716a398fdb175936e520c6cb345183), I see that the Jenkinsfile is the principal's branch one

dduportal commented:
> > I changed `${WORKSPACE}` to `${env.WORKSPACE}` after checking some documentation, let's see if this makes a difference.
> 
> good catch. However, it does not look like that your commit are considered trusted: if I click "Replay" on the build 38 (triggered by [3a357da](https://github.com/jenkins-infra/stories/commit/3a357da057716a398fdb175936e520c6cb345183)), I see that the Jenkinsfile is the principal's branch one

@krisstern do not push any more change: I'm triggering a new build with the Jenkinsfile from https://github.com/jenkins-infra/stories/commit/3a357da057716a398fdb175936e520c6cb345183

dduportal commented:
> > > I changed `${WORKSPACE}` to `${env.WORKSPACE}` after checking some documentation, let's see if this makes a difference.
> > 
> > 
> > good catch. However, it does not look like that your commit are considered trusted: if I click "Replay" on the build 38 (triggered by [3a357da](https://github.com/jenkins-infra/stories/commit/3a357da057716a398fdb175936e520c6cb345183)), I see that the Jenkinsfile is the principal's branch one
> 
> @krisstern do not push any more change: I'm triggering a new build with the Jenkinsfile from [3a357da](https://github.com/jenkins-infra/stories/commit/3a357da057716a398fdb175936e520c6cb345183)

https://ci.jenkins.io/job/Websites/job/stories/job/PR-79/39:

=> the environment variable  is set up by the pipeline with the proper value, but either it is not read by whatever Node/NPM command or it is not the correct one. Can you try this variable on your local machine please?

krisstern commented:
Hi @dduportal I need your help here if you have a few minutes to spare. I have updated the `Jenkinsfile` but it seems like the new changes are not reflected. Could you please help?

dduportal commented:
> Hi @dduportal I need your help here if you have a few minutes to spare. I have updated the `Jenkinsfile` but it seems like the new changes are not reflected. Could you please help?

https://ci.jenkins.io/job/Websites/job/stories/job/PR-79/43/

dduportal commented:
> > Hi @dduportal I need your help here if you have a few minutes to spare. I have updated the `Jenkinsfile` but it seems like the new changes are not reflected. Could you please help?
> 
> https://ci.jenkins.io/job/Websites/job/stories/job/PR-79/43/

@krisstern still the same error `NestedError: Cannot copy from `.cache/.eslintrc.json` to `/opt/build/cache/cwd  
/.cache/.eslintrc.json`: Cannot create directory `/opt/build/cache/cwd/.cache`  
: EACCES: permission denied, mkdir '/opt/build'`

=> can you try to run the same commands on your machine or inside a container based on `jenkinsciinfra/jenkins-agent-ubuntu-22.04:2.33.0` to reproduce it?

krisstern commented:
Sure, will do so later

dduportal reviewed (commented):
(side note: all occurences of `npm install` or `npm i` MUST be changed to `npm clean-install` or `npm ci`. There are no good reason to run npm install in a CI build)

dduportal commented:
Replay with 

```
pipeline {
  options {
    timeout(time: 60, unit: 'MINUTES')
    ansiColor('xterm')
    disableConcurrentBuilds(abortPrevious: true)
    buildDiscarder logRotator(artifactDaysToKeepStr: '', artifactNumToKeepStr: '', daysToKeepStr: '5', numToKeepStr: '5')
  }

  agent {
    label 'linux-arm64-docker || arm64linux'
  }

  environment {
    NODE_ENV = 'production'
    TZ = "UTC"
    NETLIFY = "true"
    // Amount of available vCPUs, to avoid OOM - https://www.gatsbyjs.com/docs/how-to/performance/resolving-out-of-memory-issues/#try-reducing-the-number-of-cores
    // https://github.com/jenkins-infra/jenkins-infra/tree/production/hieradata/clients/controller.ci.jenkins.io.yaml#L327
    GATSBY_CPU_COUNT = "4"
    // Added the below to fix permissions issue with the cache
    GATSBY_CACHE_DIR = "${env.WORKSPACE}/.gatsby-cache"
    GATSBY_INTERNAL_CACHE_DIR = "${env.WORKSPACE}/.cache"
    GATSBY_TELEMETRY_DISABLED = "1"
    NODE_OPTIONS = "--no-warnings"
  }

  stages {
    stage('Debug DDU') {
      steps {
        sh '''
         echo "$HOME"
         echo "$GATSBY_CACHE_DIR"
        '''
      }
    }
  }
}
```

shows that the variable and the syntax are good:

```
20:28:37  [Pipeline] stage
20:28:37  [Pipeline] { (Debug DDU)
20:28:38  [Pipeline] sh
20:28:37  + echo /home/jenkins
20:28:37  /home/jenkins
20:28:37  + echo /home/jenkins/agent/workspace/Websites_stories_PR-79/.gatsby-cache
20:28:37  /home/jenkins/agent/workspace/Websites_stories_PR-79/.gatsby-cache
20:28:38  [Pipeline] }
```

=> there is a profile or a setting which most probably overload the value (or it is not the only variable to specify something into `/opt`)

krisstern commented:
Used `npm ci` for installing dependencies and fixed up some `eslint` issues due to config setup, now the "Lint and Test" step is passing as well.

dduportal commented:
Pushed a commit removing the environment variable `NETLIFY`: https://github.com/jenkins-infra/stories/pull/79/commits/7f6bf2ecf99d414b22fd431bd24b01cf82bd87fc (credits to @halkeye ❤️ - https://matrix.to/#/!JLUOInpEYmxJIYXlzs:matrix.org/$MS0P293hRIirE_v8TeX7sEBUoGSscOuP7RGoiVqcGpQ?via=g4v.dev&via=gitter.im&via=matrix.org)

biru-codeastromer commented:
Very happy to see the progress here finally that ci checks passed 🥳. only one is showing failed others passed

dduportal commented:
> Pushed a commit removing the environment variable `NETLIFY`: [7f6bf2e](https://github.com/jenkins-infra/stories/commit/7f6bf2ecf99d414b22fd431bd24b01cf82bd87fc) (credits to @halkeye ❤️ - https://matrix.to/#/!JLUOInpEYmxJIYXlzs:matrix.org/$MS0P293hRIirE_v8TeX7sEBUoGSscOuP7RGoiVqcGpQ?via=g4v.dev&via=gitter.im&via=matrix.org)
> 
> Replaying in https://ci.jenkins.io/job/Websites/job/stories/job/PR-79/48/ with the proper `Jenkinsfile`

Worked: it is green in ci.jenkins.io. Replaying in infra.ci for the preview deploy step: https://infra.ci.jenkins.io/job/website-jobs/job/stories/job/PR-79/40/

biru-codeastromer commented:
It's all green now , so much happy about that finally this pr gets green 😇

biru-codeastromer commented:
<img width="1512" alt="Screenshot 2025-03-29 at 12 47 29 PM" src="https://github.com/user-attachments/assets/e636cc21-215a-403e-bdd5-647c6411b5e2" />

the deployment is up and running🚀

halkeye commented:
Btw env is consumed https://github.com/jenkins-infra/gatsby-plugin-jenkins-layout/blob/main/makeLayout.js#L83

It's probably not a big deal as netlify didn't require the icon for sponsorship, just the mention on Jenkins.io

dduportal commented:
> > Pushed a commit removing the environment variable `NETLIFY`: [7f6bf2e](https://github.com/jenkins-infra/stories/commit/7f6bf2ecf99d414b22fd431bd24b01cf82bd87fc) (credits to @halkeye ❤️ - https://matrix.to/#/!JLUOInpEYmxJIYXlzs:matrix.org/$MS0P293hRIirE_v8TeX7sEBUoGSscOuP7RGoiVqcGpQ?via=g4v.dev&via=gitter.im&via=matrix.org)
> > Replaying in https://ci.jenkins.io/job/Websites/job/stories/job/PR-79/48/ with the proper `Jenkinsfile`
> 
> Worked: it is green in ci.jenkins.io. Replaying in infra.ci for the preview deploy step: https://infra.ci.jenkins.io/job/website-jobs/job/stories/job/PR-79/40/

Now we validated it works, I thing that @halkeye 's suggestion about doing a distinct PR with only the `Jenkinsfile` would make your life easier. I'm taking care of it so we can, then, rebase this one an you will have a fully set up pipeline in autonomy
